fluent-plugin-events/lib/fluent/plugin/in_events.rb

```diff
@@ -0,0 +1,144 @@
+      config_param :secret_dir, :string, default: '/var/run/secrets/kubernetes.io/serviceaccount'
+      config_param :bearer_token_file, :string, default: nil
+      config_param :verify_ssl, :bool, default: true
+      # if `ca_file` is for an intermediate CA, or otherwise we do not have the
+      # root CA and want to trust the intermediate CA certs we do have, set this
+      # to `true` - this corresponds to the openssl s_client -partial_chain flag
+      # and X509_V_FLAG_PARTIAL_CHAIN
+      config_param :ssl_partial_chain, :bool, default: false
+      
+      # parameters for events collection
+      config_param :tag, :string, default: 'kubernetes.*'
+      config_param :namespace, :string, default: nil
+      config_param :label_selector, :string, default: nil
+      config_param :field_selector, :string, default: nil
+      
+      def configure(conf)
+        super
+        normalize_param
+        connect_kubernetes
+      end
+
+      def start
+        super
+        initialize_resource_version
+        start_watcher_thread
+        start_configmap_flush_thread
+      end
+  
+      def close
+        @watcher.each &:finish
+        super
+      end
+
+      def initialize_resource_version
+        # get or create the config map
+        begin
+          @client.public_send("get_config_map", "fluentd-config-resource-version", "sumologic").tap do |resource|
+            log.info {"Get config maps: #{resource}"}
+            version = resource.data['resource-version']
+            @resource_version = version if version
+          end
+        rescue Kubeclient::ResourceNotFoundError
+          create_config_maps
+        end
+      end
+
+      def create_config_maps
+        @resource_version = "0"
+        resource = ::Kubeclient::Resource.new
+        resource.metadata = {
+          name: "fluentd-config-resource-version",
+          namespace: "sumologic"
+        }
+        resource.data = { "resource-version": "#{@resource_version}" }
+        @client.public_send("create_config_map", resource).tap do |maps|
+          log.info {"Created config maps: #{maps}"}
+        end
+      end
+  
+      def start_watcher_thread
+        params = Hash.new
+        params[:as] = :raw
+        params[:resource_version] = @resource_version
+        params[:field_selector] = @field_selector
+        params[:label_selector] = @label_selector
+        params[:namespace] = @namespace
+
+        @watcher = @client.public_send("watch_events", params).tap do |watcher|
+          thread_create(:"watch_events") do
+            watcher.each do |entity|
+              log.debug { "Received new object from watching events" }
+              entity = JSON.parse(entity)
+              router.emit tag, Fluent::Engine.now, entity
+              @resource_version = entity['object']['metadata']['resourceVersion']
+  
+              if (!@resource_version)
+                @resource_version = 0
+                sleep(5)
```

Comment:
Maybe this one can also be configurable?

Reply:
yeah I thought about this, but since this is just a safeguard to avoid the plugin going into a hot loop when in case every event it gets has empty resource version, I thought it probably doesn't really provide much value to make it configurable for the user